sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,621 @@
+
+
+=== Background
+
+The DPC++ bindless images extension has sought to provide the flexibility of
+bindless images at the SYCL application level. Per the proposal, users would be
+able to separate image memory allocation from the image handle creation.
+Tools such as SYCLomatic would benefit from this flexibility when converting
+CUDA to SYCL code. Level Zero's support for this extension would enable
+converted code to run optimally on Intel hardware.
+
+Additionally, per the DPC++ bindless images extension, users would be able to
+create and sample images based on linearly allocated memory. A valuable use-case
+is the creation of a bindless image on already existing USM memory, preventing
+copies, and hardware sampling can be applied to that data.
+
+This proposal aims to extend Level Zero to better align with the PI API used by
+the SYCL runtime in DPC++.
+
+
+=== Specification
+
+To enable the separation of image memory allocation from image handle
+generation, we propose two new allocation functions `zeMemAllocImageExp` and
+`zeMemAllocPitchedExp`.
+
+`zeMemAllocImageExp` will allocate non-USM image memory with an implementation
+specific layout (such as tile swizzle patterns or lossless compression).
+
+Whereas, `zeMemAllocPitchedExp` will allocate USM memory in a linear layout.
+Memory allocated with this API should be freed using `zeMemFree`.
+
+Alongside these allocation functions, we propose a singular additional freeing
+function, `zeMemFreeImageExp`, to free memory allocated with
+`zeMemAllocImageExp`.
+
+We also propose a new struct `ze_device_image_bindless_exp_desc_t` that
+describes device restrictions on image properties. It should be passed to
+`ze_device_properties_t::pNext`. It contains the fields `imagePitchAlign`,
+`maxImageLinearWidth`, `maxImageLinearHeight`, and `maxImageLinearPitch`.
+
+== Definitions
+
+```cpp
+typedef uint64_t ze_image_mem_handle_exp_t;
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Describes memory to be allocated on device intended for images with
+///        a device specific memory layout
+///
+/// @details
+///     - This structure is passed to:
+///       + ::zeMemAllocImageExp
+///       + ::ze_image_desc_t.pNext
+///         - When this structure is passed to an image descriptor,
+///           ::zeImageCreate must not allocate any additional device memory for
+///           the created image, only return a handle.
+///         - Additionally, when ::zeImageDestroy is called on an image created
+///           from a ::ze_image_mem_alloc_exp_desc_t allocation,
+///           ::zeImageDestroy must not free the memory associated with the
+///           image handle.
+///
+typedef struct _ze_image_mem_alloc_exp_desc_t {
+    ze_structure_type_t stype;      // [in]
+    void *pNext;                    // [in,out][optional]
+    size_t widthInBytes;            // [in]
+    size_t height;                  // [in][optional]
```

Comment:
depth is usually used for 3D images -are 3D images supported ?

does widthInBytes already include padding to row pitch alignment ?

is this the formula to calculate total memory size : widthInBytes * height * depth * elementSizeByte ?

Reply:
Yes. 3D images are supported.

Yes, `widthInBytes` already includes padding and alignment.

No, the formula for total memory size would be 
```
widthInBytes * height * depth
```
and the formula for number of elements would be
```
(widthInBytes / elementSizeByte) * height * depth
```

Reply:
so i think slice padding ( if exists, for a case where slice size != rowpitch * elementSize * height  ) is not included

Reply:
Hi @HoppeMateusz ,

You'll have to forgive me, I've answered your original question wrong. To go back on it, `widthInBytes` **does not** include padding/alignment, and is merely the requested allocation width (in bytes). I'm very sorry for any confusion from this.

The relevant formulas would be
```c++
total_memory_size_2D = rowPitch * height * depth
total_memory_size_3D = rowPitch * slicePitch * depth
``` 

To answer your questions about support for 3D images, I hope the latest revision clears up any issues as it has some changes around this part of the spec. To summarize the changes, the row pitch and slice pitch should be handled by the proposed implementation-specific `ze_image_mem_handle_exp_t`, whereas width, height, depth, element size, etc. should be piggy-backed off the existing `ze_image_desc_t`.